app/setup/index.js

```diff
@@ -0,0 +1,259 @@
+        this.updateStateWithEvent = this.updateStateWithEvent.bind(this);
+        this.verifyAuthDone = this.verifyAuthDone.bind(this);
+    }
+
+    buildOauthUrl() {
+        const oauthClientId = 'isFcew9naom2f1khSiMeAtzuOvHXHuLwhPsM7oPt';
+        const isOnPrem = this.state.serverType === ONPREM;
+        const plotlyDomain = isOnPrem ? this.state.domain : 'plot.ly';
+        const redirect_uri = baseUrl().replace('/setup', '');
+        return (
+            `https://${plotlyDomain}/o/authorize/?response_type=token&` +
+            `client_id=${oauthClientId}&` +
+            // TODO: Is there a better way? On premise has the `external-data-connector`
+            // route so cant use window.location.origin.
+            `redirect_uri=${redirect_uri}/oauth2/callback`
+        );
+    }
+
+    chooseServerType(serverType) {
+        this.setState({serverType});
+    }
+
+    verifyAuthDone() {
+        const {username} = this.state;
+        console.log(`baseUrl() ${baseUrl()}`);
+        fetch(`${baseUrl()}/hasauth`, {
+            method: 'POST',
+            headers: {
+                'Accept': 'application/json',
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify({
+                username
+            })
+        }).then(res => res.json().then(json => {
+            if (res.status !== 200) {
+                this.setState({status: 'failure', errorMessage: json.error.message});
+            }
+            this.setState({loggedIn: json.hasAuth});
+        }));
+        // TODO: More error handling here? Probably need a catch here.
+    }
+
+    createCertificates() {
+        const {username} = this.state;
+        fetch(`${baseUrl()}/hasauth`, {
+            method: 'POST',
+            headers: {
+                'Accept': 'application/json',
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify({
+                username
+            })
+        }).then(res => res.json().then(json => {
+            if (res.status !== 200) {
+                this.setState({status: 'failure', errorMessage: json.error.message});
+            }
+            this.setState({loggedIn: json.hasAuth});
+        }));
+    }
+
+    setTimeoutAuthDone () {
+        const {username} = this.state;
+        const checkAuth = setInterval(() => {
+            // TODO: This is not very clear for a message. Show them a link to the
+            // maybe in case they closed it or want to try again?
+            this.setState({
+                errorMessage: `We\'re waiting for authorization of [${username}].`});
+            this.verifyAuthDone();
+            if (this.state.loggedIn) {
+                // TODO: Is there a better way? On premise has the `external-data-connector`
+                // route so cant use window.location.origin.
+                clearInterval(checkAuth);
+                const appRoute = baseUrl().replace('/setup', '/');
```

Comment:
if we want users to always go through this flow, should this login screen be the default URL and then `connections` or something be the new main screen?